File: api/src/main/java/com/wbrawner/budgetserver/budget/Budget.java
package com.wbrawner.budgetserver.budget;

import com.wbrawner.budgetserver.category.Category;
import com.wbrawner.budgetserver.transaction.Transaction;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;
import java.util.TreeSet;

@Entity
public class Budget {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;
    private String name;
    private String description;
    private String currencyCode;
    @OneToMany(mappedBy = "budget")
    private final Set<Transaction> transactions = new TreeSet<>();
    @OneToMany(mappedBy = "budget")
    private final Set<Category> categories = new TreeSet<>();
    @OneToMany(mappedBy = "budget")
    private final Set<Transaction> users = new HashSet<>();

    public Budget() {}

    public Budget(String name, String description) {
        this(name, description, "USD");
    }

    public Budget(String name, String description, String currencyCode) {
        this.name = name;
        this.description = description;
        this.currencyCode = currencyCode;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getCurrencyCode() {
        return currencyCode;
    }

    public void setCurrencyCode(String currencyCode) {
        this.currencyCode = currencyCode;
    }

    public Set<Transaction> getTransactions() {
        return transactions;
    }

    public Set<Category> getCategories() {
        return categories;
    }

    public Set<Transaction> getUsers() {
        return users;
    }
}

File: api/src/main/kotlin/com/wbrawner/budgetserver/budget/Budget.kt
package com.wbrawner.budgetserver.budget

import com.wbrawner.budgetserver.category.Category
import com.wbrawner.budgetserver.permission.UserPermission
import com.wbrawner.budgetserver.permission.UserPermissionRequest
import com.wbrawner.budgetserver.permission.UserPermissionResponse
import com.wbrawner.budgetserver.randomString
import com.wbrawner.budgetserver.transaction.Transaction
import java.util.*
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.OneToMany

@Entity
data class Budget(
    @Id
    var id: String = randomString(),
    var name: String? = null,
    var description: String? = null,
    var currencyCode: String? = "USD",
    @OneToMany(mappedBy = "budget")
    val transactions: Set<Transaction> = TreeSet(),
    @OneToMany(mappedBy = "budget")
    val categories: Set<Category> = TreeSet(),
    @OneToMany(mappedBy = "budget")
    val users: Set<Transaction> = HashSet()
)

data class BudgetRequest(
    val name: String = "",
    val description: String = "",
    val users: Set<UserPermissionRequest> = emptySet()
)

data class BudgetResponse(
    val id: String,
    val name: String?,
    val description: String?,
    private val users: List<UserPermissionResponse>
) {
    constructor(budget: Budget, users: List<UserPermission>) : this(
        Objects.requireNonNull<String>(budget.id),
        budget.name,
        budget.description,
        users.map { userPermission: UserPermission -> UserPermissionResponse(userPermission) }
    )
}

data class BudgetBalanceResponse(val id: String, val balance: Long)